Draft: Dimension styles

carlopav opened the pull request:
This PR provide objects to store annotation styles inside the document, and modify the annotation objects to store the link to their stile.

To be merged after PR #3102

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [ ] Branch rebased on latest master `git pull --rebase upstream master`
- [ ] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

vocx-fc commented on src/Mod/Draft/draftviewproviders/view_style_dimension.py:
```diff
@@ -0,0 +1,230 @@
+        """
+        Apply the style to the related dimensions
+        """
+        from draftutils import gui_utils
```
This import should be at the top of the file, after `draftutils.utils`, not inside the function.

vocx-fc commented on src/Mod/Draft/draftguitools/gui_dimensionstyle.py:
```diff
@@ -32,18 +32,78 @@
     """
+    def __init__(self):
+        super().__init__()
+        self.command_name = "DimensionStyle"
```
This should be
```py
def __init__(self):
    super().__init__(name="Dimension style")
```
The parent class, GuiCommandSimplest, sets up the command_name attribute.

vocx-fc commented on src/Mod/Draft/draftobjects/dimension.py:
```diff
@@ -0,0 +1,235 @@
+
+
+    def execute(self, obj):
+        import DraftGeomUtils
```
This should be imported at the top of the file.

vocx-fc commented on src/Mod/Draft/draftobjects/dimension.py:
```diff
@@ -0,0 +1,235 @@
+class LinearDimension(DimensionBase):
+    """The Draft Dimension object"""
+    def __init__(self, obj):
+        DimensionBase.__init__(self,obj,"Dimension")
```
The better way is to call `super().__init__(self,obj,"Dimension")`.

We prefer to avoid repeating the parent class in case it changes.

vocx-fc commented on src/Mod/Draft/draftobjects/dimensionstyle.py:
```diff
@@ -28,13 +28,12 @@
 if App.GuiUp:
     import FreeCADGui as Gui
-    from draftviewproviders.view_style_dimension import ViewProviderDraftDimensionStyle
+    from draftviewproviders.view_dimensionstyle import ViewProviderDraftDimensionStyle
```
I don't think we need to guard the view providers because they won't be executed by the import, so there is no risk of causing an error. I mean, it doesn't hurt at all to guard them like this, I just think it looks a bit ugly.

We have to guard `FreeCADGui` because if we don't we will definitely cause an error in console mode.

vocx-fc replied:
About this, I think some Travis tests are failing when we import the view provider module because inside this module `FreeCADGui` is not guarded.

So inside `view_dimensionstyle.py` and all those new view providers.
```py
if App.GuiUp:
    import FreeCADGui as Gui
```

vocx-fc commented on src/Mod/Draft/draftviewproviders/view_dimension.py:
```diff
@@ -0,0 +1,549 @@
+        self.node3d.removeChild(self.marks)
+
+    def draw_dim_arrows(self, vobj):
+        from pivy import coin
```
We shouldn't be importing coin inside functions. These should be at the top.

vocx-fc commented on src/Mod/Draft/Draft.py:
```diff
@@ -252,79 +252,14 @@ def makeRectangle(length, height, placement=None, face=None, support=None):
+from draftviewproviders.view_dimension import ViewProviderLinearDimension
+_ViewProviderDimension = ViewProviderLinearDimension
 
     return obj
```
@carlopav You forgot to remove this `return obj` statement, and when the entire `Draft.py` is imported, it causes a syntax error. The Travis tests show this and fail because of it.

vocx-fc commented:
Just another note, the Travis tests fail because when other workbenches import `Draft.py`, the `draftobjects.dimension` module is not found. The problem is that this module was not added to the `CMakeLists.txt` file. So, the new files `dimension.py`, `view_dimension.py`, etc. must be added to the appropriate variable in `CMakeLists.txt`, then they will be installed correctly in Travis, and hopefully the tests will pass.

vocx-fc commented on src/Mod/Draft/draftobjects/dimension.py:
```diff
@@ -0,0 +1,373 @@
+        elif p3 == "diameter":
+            #l.append((p1,"Diameter"))
+            if App.GuiUp:
+                obj.ViewObject.Override = "Ø $dim"
```
The Travis tests are failing in this line because of the unicode character `"Ø $dim"` . You have to add the codification at the beginning of the file. `# -*- coding: utf8 -*-`

carlopav replied:
thanks for looking into this. I'll add it right now.

carlopav replied:
shouldn't I add it to every file?

vocx-fc replied:
I wouldn't add it to every file, only to the one reported by our tests. This is actually only required in Python 2. But if it causes this error in Travis, it may mask other errors elsewhere, so we try to have no errors at all when making a pull request.

vocx-fc commented on src/Mod/Draft/draftobjects/label.py:
```diff
@@ -0,0 +1,233 @@
+            elif obj.LabelType == "Area":
+                if hasattr(obj.Target[0],'Shape'):
+                    if hasattr(obj.Target[0].Shape,"Area"):
+                        obj.Text = [App.Units.Quantity(obj.Target[0].Shape.Area,App.Units.Area).UserString.replace("^2","²")]
```
This small 2 is also causing codification problems. `# -*- coding: utf8 -*-`

carlopav replied:
thx for spotting it!

carlopav commented:
@vocx-fc any idea why the build fail on linux? (if I can read it correctly)
https://travis-ci.org/github/FreeCAD/FreeCAD/builds/666987452?utm_medium=notification&utm_source=github_status

vocx-fc commented on src/Mod/Draft/Draft.py:
```diff
@@ -170,6 +170,81 @@
+    from draftviewproviders.view_dimension import ViewProviderAngularDimension
+    _ViewProviderDimension = ViewProviderLinearDimension
+
+_ViewProviderAngularDimension = ViewProviderAngularDimension
```
In Travis the gui doesn't exist. So this view provider also has to be inside the `if gui:` so it is skipped.

carlopav replied:
good catch!

carlopav commented:
Bingo! thx @vocx-fc

carlopav commented:
@yorikvanhavre @vocx-fc 
I think we can assume the PR is ready, I'm removing the Draft mark.
Let me know when I have to rebase it to upstream.

vocx-fc commented:
This branch follows from the Draft status bar, right? So, it should be merged after it. Which pull request is it?

carlopav commented:
I suggest to first merge this one and later https://github.com/FreeCAD/FreeCAD/pull/3228.

yorikvanhavre commented:
Ok now this PR is next in the queue @carlopav ! Can you rebase it and solve the conflicts?

carlopav commented:
@yorikvanhavre are you sure you want this to be merged? 
We will have documents with this dimension style object and if we want to get rid of it later this will be a big problem, no?

yorikvanhavre commented:
Hmm you have a point there... Generally speaking I'm fine with experimentation while we are in "development branch" mode, but indeed this could be annoying. Splitting seems a good idea!

carlopav commented:
ok, so let's go in this direction. Feature freeze could be near and it would be a bad thing to release 0.19 with this feature.
I'll rebase it tomorrow i think.

carlopav commented:
@yorikvanhavre The branch has been rebased, and I added on top a commit to remove the Draft Dimension Style object. So at the moment the title is a bit inappropriate.
What the branch introduces is:
- splitting of annotation objects
- derive them from common base classes
- added ScaleMultiplier property to all annotations
- added a statusbar scale widget to select scale while drawing.
If it compiles also on travis, (compiling was successful on my machine), I think you can evaluate it for merging!

vocx-fc commented on src/Mod/Draft/draftobjects/dimension.py:
```diff
@@ -0,0 +1,370 @@
+
+    def __init__(self, obj):
+
+        super().__init__(obj,"AngularDimension")
```
This is failing the Travis tests for Python 2. In this version the syntax is `super(DimensionBase, self)`. It probably should be corrected if this is going to be released in Python 2.

carlopav replied:
It's just good if I substitute it?

yorikvanhavre replied:
An easy way to fix this (not sure about how pythonic that is...) woud be instead of calling super(), calling explicitely DimensionBase.__init__(obj,"AngularDimension") I think that works equally for both python versions

carlopav commented:
oops, i did not realize the build failed... Ok I'm trying what you suggest yorik, hope this is the only thing that make it fail :)

carlopav commented:
@yorikvanhavre @vocx-fc 
changed. let's see if it compiles correctly now!

carlopav commented:
@yorikvanhavre good catch, I guess we can keep it like this for the moment!
@vocx-fc is it ok for you?

vocx-fc commented:
Well, using the parent class explicitly is not Pythonic because if we move the class we have to change all references to that class, so that's why `super` is better.

Use the parent class `DimensionBase` explicitly.
```py
DimensionBase.__init__(self, obj,"AngularDimension")
```

Use whichever is the parent class of our current class, `AngularDimension`. This is better.
```py
super(AngularDimension, self).__init__(obj, "AngularDimension")
```

See the last commit in #3344.

carlopav commented:
this was the first attempt but resulted in a failed compile, maybe I did something wrong and I did not realize it...
I'll make another attempt!

carlopav commented:
done, finger crossed!

carlopav commented:
no way... it fails again... 
@vocx-fc I really can't understand what i do wrong, and I can't debug it because it compiles fine on my pc...
Should we go back to `DimensionBase.__init__(self, obj,"AngularDimension")`

vocx-fc commented:
The build is failing in Python 2 with this error message, `TypeError: super() argument 1 must be type, not classobj`. This happens because `super()` only works with "new style" classes, and not "old style" classes.

In Python 3 all classes are new style so this is not a problem. In Python 2, classes must be explicitly defined as new style.

In Python 2 all classes that don't have a parent class are of type `classobj` and not of type `type`. This can be solved by declaring the parent class as
```py
class Parent(object):
    ...
```

Where `object` is basic built-in type.

Another way is to leave the class declaration without a parent class but using the `__metaclass__` variable.

```py
__metaclass__ = type  # to support Python 2 use of `super()`
class Parent:
    ...
```

I was actually not aware of this until now. More information here, [super() fails with error: TypeError “argument 1 must be type](https://stackoverflow.com/questions/1713038/super-fails-with-error-typeerror-argument-1-must-be-type-not-classobj-when).

carlopav commented:
oh, thanks a lot for digging into this...
So it's just a matter of declaring `class DraftAnnotation(object):` and `class ViewProviderDraftAnnotation(object):`, because all the other classes inherits from those.
Strange that this did not happen with the code you were splitting, but just with those draft_objects...

vocx-fc commented:
The code that I'm modifying consists of graphical tools. These are used from the graphical interface, so there are no unit tests for them. Travis runs the unit tests without the graphical interface, so it can only catch non-GUI errors in the object code, it will not catch errors with any of the graphical tools.

carlopav commented:
Yes, right! that's it. OK, later i'll modify the code again and integrate your two comments above!
thanks :)

carlopav commented:
@vocx-fc thanks! it worked. (i didn't realize also the base object for bim assembly experimentations that realthunder provided had the `(object)` too)
@yorikvanhavre here we are!

yorikvanhavre commented:
Victory!!

carlopav commented:
Wow! A big thank you to everyone! :D

vocx-fc commented:
I just noticed that the image files `Draft_Annotation_Style.svg`, `Draft_Dimension_Style.svg`, and `Draft_Dimension_Style_Tree.svg` aren't shown in GitHub. It says that the files are invalid. This probably means that the files have some non-standard Inskcape tags. So, when you save icons, make sure they are saved as Plain SVG.

vocx-fc commented:
Also, I noticed that that the `GuiCommandSimplest` and `GuiCommandNeedsSelection` classes in `gui_base.py` were added twice.

Once they were added in this branch (530e8b6c58d), and the second time they were added in my branch #3182 (017226e8037). Obviously this is not ideal, so I rebased my branch #3291 and added a new commit to remove the repeated classes.

vocx-fc commented:
Also, I integrated the small change made to the Draft_Label tool, in ff3bdf86bc1df, to my branch #3299.

carlopav commented:
My bad, those Icons (`Draft_Annotation_Style.svg`, `Draft_Dimension_Style.svg`, `Draft_Dimension_Style_Tree.svg`) are obsolete, i was sure I had removed them... 
They were used for the DimensionStyle object that we decided to discard.
 I'll make another PR to remove them. Thanks also for the other things spotted :)

vocx-fc commented:
About the icons, it seems you only added one, however, I checked your partial commits and that's when I noticed the error displaying all of them. Nevertheless, I'm unsure if this is really a problem from your side. Maybe something is wrong with GitHub at the moment because I can't see any icons.

For example, #3060 introduced many icons and they were visible without problems then, but now they aren't displayed. I've read that GitHub has been having a few issues the last few weeks, so that may be a cause.

carlopav commented:
Yes, just one of the icons is still present in the code. I think I just forget about it... I'll delete it during the weekend!

vocx-fc commented:
Just one more thing Carlo. Can you check why the size of the text is now different? Use the test script.
```py
import drafttests.draft_test_objects as dt
dt.create_test_file()
```

Before your changes, all text would fit correctly in the rectangle, now it's too big. Maybe the sizes need to be set to a different value in the test script.

carlopav commented:
I think this is because of the scale multiplier property: when you create a dimension it automatically takes the ScaleMultiplier property according to the statusbar scale.
If you set this property to 1, they should look good. Also if you set the scale 1:1 in the statusbar, they will be created correctly.
@yorikvanhavre Anyway I spotted a bug and PR the fix: https://github.com/FreeCAD/FreeCAD/pull/3357

vocx-fc commented:
Yes, I figured that much, the issue is the default scale multiplier is too large. In that case, you should set the scale multiplier to 1 by default to keep the same presentation as before. Otherwise new users will see their texts changed and won't know the reason.

carlopav commented:
well, not exactly: this affect only new created objects. 
So if the user had them from before, or close and open the document, they will be displayed correctly.
The problem with the test is that it creates new objects...perhaps we can modify the make_dimension and make_text function to accept also this parameter... ?

vocx-fc commented:
Yes, but why should the default scale be anything but 1? That's more or less the issue. The test script is supposed to produce the same result every time it is run; if it doesn't it means something is changing the behavior. In this case, I think the natural solution is to set the scale to 1 by default, not to modify the make functions. Of course, the make functions can also be improved, but that's a separate discussion.

carlopav commented:
The default is in fact set to 1, but it is stored inside FreeCAD parameters:
`        param = App.ParamGet("User parameter:BaseApp/Preferences/Mod/Draft")
        annotation_scale = param.GetFloat("DraftAnnotationScale", 1.0)
        if annotation_scale != 0:
            vobj.ScaleMultiplier = 1 / annotation_scale`
this is the code involved.
The point is that you did set it to something else last time, it keeps the different scale, as all other parameters do in Draft Annotation since before I started editing them.
So, this is a good point. Are those supposed to be Document parameters or General parameters? Is this the right question to ask?

vocx-fc commented:
I set it to something else myself? Huh? Well, I wouldn't know. Maybe at some point I tested your branch and that value was written at that point.

I think keeping them as Draft parameters is fine.

This is a similar discussion as with your annotation styles concept. If you move the file to another system, it should display the same, and not change the text arbitrarily because of the different setup. This works correctly right now, because you save the scale as a property, so it should work fine. As long as only new objects are affected, there shouldn't be a problem.
